File: src/main/java/com/dozycoffee/wms/warehouse/domain/valueobject/Coordinate.java
package com.dozycoffee.wms.warehouse.domain.valueobject;

import com.dozycoffee.wms.warehouse.domain.exception.InvalidCoordinateException;
import java.math.BigDecimal;

public record Coordinate(
        BigDecimal latitude,
        BigDecimal longitude
) {
    private static final BigDecimal MIN_LATITUDE = BigDecimal.valueOf(-90);
    private static final BigDecimal MAX_LATITUDE = BigDecimal.valueOf(90);
    private static final BigDecimal MIN_LONGITUDE = BigDecimal.valueOf(-180);
    private static final BigDecimal MAX_LONGITUDE = BigDecimal.valueOf(180);

    public Coordinate {
        if (latitude == null || longitude == null) {
            throw new InvalidCoordinateException();
        }
        if (latitude.compareTo(MIN_LATITUDE) < 0 || latitude.compareTo(MAX_LATITUDE) > 0) {
            throw new InvalidCoordinateException();
        }
        if (longitude.compareTo(MIN_LONGITUDE) < 0 || longitude.compareTo(MAX_LONGITUDE) > 0) {
            throw new InvalidCoordinateException();
        }
    }

    public static Coordinate of(double latitude, double longitude) {
        return new Coordinate(BigDecimal.valueOf(latitude), BigDecimal.valueOf(longitude));
    }
}

File: src/main/kotlin/com/dozycoffee/wms/warehouse/domain/valueobject/Coordinate.kt
package com.dozycoffee.wms.warehouse.domain.valueobject

import com.dozycoffee.wms.warehouse.domain.exception.InvalidCoordinateException
import java.math.BigDecimal

@ConsistentCopyVisibility
data class Coordinate private constructor(
    val latitude: BigDecimal,
    val longitude: BigDecimal
) {
    companion object {
        private val MIN_LATITUDE: BigDecimal = BigDecimal.valueOf(-90)
        private val MAX_LATITUDE: BigDecimal = BigDecimal.valueOf(90)
        private val MIN_LONGITUDE: BigDecimal = BigDecimal.valueOf(-180)
        private val MAX_LONGITUDE: BigDecimal = BigDecimal.valueOf(180)

        fun of(latitude: BigDecimal?, longitude: BigDecimal?): Coordinate {
            if (latitude == null || longitude == null) {
                throw InvalidCoordinateException()
            }
            if (latitude < MIN_LATITUDE || latitude > MAX_LATITUDE) {
                throw InvalidCoordinateException()
            }
            if (longitude < MIN_LONGITUDE || longitude > MAX_LONGITUDE) {
                throw InvalidCoordinateException()
            }
            return Coordinate(latitude, longitude)
        }

        fun of(latitude: Double, longitude: Double): Coordinate {
            return of(BigDecimal.valueOf(latitude), BigDecimal.valueOf(longitude))
        }
    }
}
